### PR TITLE
bridge: cleanup ports from host after pod removal

### PR DESCRIPTION
**Pull request by phoracek**

### CODE REVIEW

File: pkg/network/bridge/plugin.go [hunk 1]
```diff
@@ -299,6 +301,8 @@ func attachPodToBridge(bridgeName, nicName string, containerPid int) error {
 		return err
 	}
 
+	nbdp.interfaceCleaner.Register(nicName, containerPid)
```
Comment (yuvalif) [inline pkg/network/bridge/plugin.go]: you should first wait for the cleaner thread to start. though it is unlikely, it may be that the cleaner did not start yet

File: pkg/network/bridge/interfacecleaner.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+		case event := <-watcher.Events:
+			if event.Op&fsnotify.Remove == fsnotify.Remove {
+				ic.mutex.Lock()
+
+				pid, err := strconv.Atoi(event.Name)
+				if err != nil {
+					panic(err)
+				}
+
+				peer, err := netlink.LinkByName(ic.hostInterfaceByPID[pid])
```
Comment (yuvalif) [inline pkg/network/bridge/interfacecleaner.go]: probably better to check if the pid exists in the map. if its not there, we can probably just ignore. it could be that some other process ended, unrelated to the allocated interface?

Comment (yuvalif) [reply pkg/network/bridge/interfacecleaner.go]: also, if you cannot get the link, IMO, don't panic, just log warning and continue - since this is a cleanup operation, its ok if the resource is already cleaned up

File: pkg/network/bridge/interfacecleaner.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+		panic(err)
+	}
+	defer watcher.Close()
+
+	err = watcher.Add("/proc")
+	if err != nil {
+		panic(err)
+	}
+
+	for {
```
Comment (yuvalif) [inline pkg/network/bridge/interfacecleaner.go]: is there a graceful shutdown mechanism for infinite loop threads?

Comment (phoracek) [reply pkg/network/bridge/interfacecleaner.go]: We can pass a stop channel, but as for now, there is no one that would call it. We can add SIGNAL handling later.

File: pkg/network/bridge/interfacecleaner.go (hunk 2 above)
Comment (phoracek) [reply pkg/network/bridge/interfacecleaner.go]: Done

File: pkg/network/bridge/plugin.go (hunk 1 above)
Comment (phoracek) [reply pkg/network/bridge/plugin.go]: Done

Comment (yuvalif) [reply pkg/network/bridge/plugin.go]: :+1:

**Comment by phoracek** [comment]
@yuvalif I had to change it, we cannot use fsnotify in /proc. Instead I check running processes every 10 seconds now and if monitored ones are gone, do the cleanup.

File: pkg/network/bridge/cleaner.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+				c.interfaceByPID[interfaceAndPID.PID] = interfaceAndPID.InterfaceName
+				c.interfaceByPIDMutex.Unlock()
+			}
+		}
+	}()
+
+	// Repeatedly check whether there are some orphan interfaces and remove them.
+	for {
+		// List all host's processes.
+		pids := make(map[int]bool)
```
Comment (yuvalif) [inline pkg/network/bridge/cleaner.go]: this, and the for loop in line 67 could take a while, if there are many processes on the host (while only a handful of them are containers). could you please look into this: https://stackoverflow.com/questions/38153350/how-to-properly-wait-for-an-event-process-to-finish-not-being-the-parent
for getting an event when a process ends (in go)

File: pkg/netlink/monitor.go [hunk 5]
```diff
@@ -0,0 +1,159 @@
+package netlink
```
Comment (yuvalif) [inline pkg/netlink/monitor.go:1]: why did you add that to their package?
should probably be internal - just in case they will add it late.
unrelated, you may submit a PR to the netlink repo for adding this?

**Review by yuvalif — APPROVED** [review]

File: pkg/netlink/monitor.go (hunk 5 above)
Comment (phoracek) [reply pkg/netlink/monitor.go:1]: I opened a card tracking this.

**Comment by phoracek** [comment]
Verified, merging.
